content/guides/nextjs/github-open-authoring/configure-custom-app.md

```diff
@@ -0,0 +1,111 @@
+
+**pages/\_app.tsx**
+
+```tsx
+import App from 'next/app'
+import { TinaCMS, TinaProvider } from 'tinacms'
+import {
+  useGithubEditing,
+  GithubClient,
+  TinacmsGithubProvider,
+} from 'react-tinacms-github'
+
+
+export default class Site extends App {
+  cms: TinaCMS
+
+  constructor(props) {
+    super(props)
+    /**
+     * 1. Create the TinaCMS instance
+     */
+    this.cms = new TinaCMS({
+      apis: {
+        /**
+         * 2. Register the GithubClient
+         */
+        github: new GithubClient({
+          proxy: '/api/proxy-github',
+          authCallbackRoute: '/api/create-github-access-token',
+          clientId: process.env.GITHUB_CLIENT_ID,
+          baseRepoFullName: process.env.REPO_FULL_NAME // e.g: tinacms/tinacms.org,
+        })
+      },
+      /**
+       * 3. Make sure the Sidebar & Toolbar are
+       *    hidden unless we're in Preview/Edit Mode
+       */
+      sidebar: {
+        hidden: !props.pageProps.preview,
+      },
+      toolbar: {
+        hidden: !props.pageProps.preview,
+      },
+    })
+  }
+
+  render() {
+    const { Component, pageProps } = this.props
+    return (
+      /**
+       * 4. Wrap the page Component with the Tina and Github providers
+       */
+      <TinaProvider cms={this.cms}>
+        <TinacmsGithubProvider
+          editMode={pageProps.preview}
+          enterEditMode={enterEditMode}
+          exitEditMode={exitEditMode}
+          error={pageProps.error}
+        >
+          {/**
+           * 5. Add a button for entering Preview/Edit Mode
+           */}
+          <EditLink editMode={pageProps.preview} />
+          <Component {...pageProps} />
+        </TinacmsGithubProvider>
+      </TinaProvider>
+    )
+  }
+}
+
+const enterEditMode = () => {
+  return fetch(`/api/preview`).then(() => {
+    window.location.href = window.location.pathname
+  })
+}
+
+const exitEditMode = () => {
+  return fetch(`/api/reset-preview`).then(() => {
+    window.location.reload()
+  })
+}
+
+export interface EditLinkProps {
+  editMode: boolean
+}
+
+export const EditLink = ({ editMode }: EditLinkProps) => {
+  const github = useGithubEditing()
+
+  return (
+    <button onClick={editMode ? github.exitEditMode : github.enterEditMode}>
+      {editMode ? 'Exit Edit Mode' : 'Edit This Site'}
+    </button>
+  )
+}
+```
+
+> _Note:_ For brevity, the example above configures many steps in a single file, but **a few components can be configured in different places**. For example you could put the `EditLink` in a Layout component, or set up the Github Provider only on certain pages.
+
+If you restart the dev server, you should see a **button in the top left-hand corner** that says, "Edit This Site". If you click it, you'll be prompted to authenticate with GitHub, but we still have some steps to get this working. Let's set up up the backend API.
```

Comment:
This is no longer true because the GithubClient freaks out if you're missing the environment variables. Maybe we shouldn't throw this early?

Also got to remember to update that URL

<img width="710" alt="image" src="https://user-images.githubusercontent.com/824015/80245363-f5515b00-8640-11ea-86fa-9329d2c53507.png">

Reply:
good call, I'll do a sweep for outdated urls

Reply:
Updated the url [here](https://github.com/tinacms/tinacms/pull/1058)

Reply:
I updated the paragraph but yeah it seems like now folks won't get a 'milestone' until the OAuth setup step...might not be ideal but let's just see if it becomes an issue

Reply:
It kind of makes me think it was a superficial milestone. You got a UI and it _looks_ like things work but they don't actually...

What if setting up the OAuth App was the first step?

Right now we setup a UI and then create the thing it connects too, which seems kind of backwards. If we started by setting up the OAuth App and the environment variables then we can explain right away what's they're for and be like "Now we'll setup the interface that let's people authenticate using this oauth app". Doing it this way would make sure the user would never have non-functioning UI.

Reply:
Hmmmm....I could be down with that!

Reply:
Updated, after reading through this order, I like it better